### PR TITLE
feat: Semrush Elements owned-urls endpoint | LLMO-6086

### PR DESCRIPTION
**Pull request by calvarezg**
## Summary
Adds the URL Inspector "Your cited URLs" table (`owned-urls`) to the Serenity Semrush-Elements wrapper, mirroring the merged cited-domains endpoint. Hybrid: per-URL citations + weekly trends from Semrush; agentic/referral traffic from Adobe Postgres.

`GET /v2/orgs/:spaceCatId/brands/:brandId/serenity/brand-presence/url-inspector/owned-urls`

## Changes
- `definitions/owned-urls.js` — stats-per-url (`9af5ed83`) + trend (`afb2e5d3`) payload builders + transform (owned filtered client-side, per-project to stay under the 50k row cap, weekly citations grouped by URL).
- `elements-service.js` — `getOwnedUrls` (per-project parallel fetch) + `getOwnedUrlProjects`.
- `owned-urls-traffic.js` — best-effort Postgres join (agentic/referral by `(site_id, url_path)`) via `rpc_url_inspector_owned_urls_traffic`; degrades to `0`/`[]`.
- `controllers/elements.js` — `listOwnedUrls`: brand auth, required+validated dates, region→project, optional brand-scoped `siteId`, client-side pagination, `cachedOk`.
- Route + `facs-capabilities` (`llmo/can_view`) + `required-capabilities` (`brand:read`) + route-registry test + docs.

## Dependency
Traffic data needs the RPC from adobe/mysticat-data-service#779 (LLMO-6086) deployed to the target env. Until then agentic/referral return `0`/`[]` (graceful); citations render from Semrush regardless.

## Test plan
- [x] `npm run lint` clean; route-registry / FACS / required-capabilities drift + existing elements tests (106) pass.
- [x] Transform + traffic-merge logic verified against live-probe-shaped data (owned filter, cross-project dedup, ISO-week grouping, graceful empty traffic).
- [x] Element shapes/payloads verified against live Semrush (stg).
- [ ] Reviewer: confirm final JSON via curl on a running local server (localhost:3002).

## API specification
OpenAPI intentionally skipped — consistent with the sibling Serenity elements endpoints (filter-dimensions / weeks / prompts / cited-domains), none of which have specs. Endpoint is implemented (no 501 stub).

## Related Issues
https://jira.corp.adobe.com/browse/LLMO-6086

## Deferred follow-ups
- Retrofit `cachedOk` (private, 2h) onto the sibling Serenity elements endpoints (`filter-dimensions`, `weeks`, `brand-presence/prompts`, `cited-domains`), which currently use plain `ok()`. This PR keeps `cachedOk` on `owned-urls` (part of the agreed caching approach); the follow-up aligns the siblings rather than dropping the better pattern. Raised by MysticatBot review.

## Also in this PR (per Vivek)
Fixes a pre-existing model-translation bug in the **filter-dimensions** builders (`brands.js`, `topics.js`): they used `ELEMENT_MODELS.includes(model) ? model : DEFAULT`, which does **no** SpaceCat→Semrush translation, so UI platform codes (`copilot`/`openai`/`gemini`/`chatgpt`) silently fell back to `search-gpt`. Switched both to the shared `resolveElementModel()` (also accepting the `platform` alias) — matching `weeks`/`cited-domains`/`prompts`/`owned-urls`. Added translation unit tests. Not originally in this workstream; folded in at Vivek's request.

### CODE REVIEW

File: src/support/elements/elements-service.js [hunk 1]
```diff
@@ -158,6 +161,73 @@ export function createElementsService(transport) {
...
+    },
+
+    /**
+     * Fetches the URL Inspector Owned URLs table (citations + weekly trends) from
+     * Semrush, backed by Stats-per-URL (9af5ed83) + URL trend (afb2e5d3). Fans out
+     * per project (region) — each project stays under the 50k-row cap and carries
+     * its region — fetching both elements in parallel, then merges to the legacy
+     * shape. Returns the FULL owned list sorted by citations desc; the controller
+     * paginates client-side and joins agentic/referral traffic for the page.
+     *
```
Comment (MysticatBot) [inline src/support/elements/elements-service.js:197]: issue (blocking): **Unbounded `Promise.all` fan-out across all projects.**

The `getOwnedUrls` method fires `Promise.all(scopes.map(...))` with no concurrency limit. Each scope spawns 2 parallel HTTP calls (stats + trend), so a brand with 15 markets triggers 30 simultaneous Semrush requests on a single user call.

The controller already defines `mapWithConcurrency` capped at `FANOUT_CONCURRENCY = 8` for exactly this pattern. Use it here (or extract it to a shared utility) to cap the outer fan-out.

```js
// Instead of:
const projectResults = await Promise.all(scopes.map(async ({ region, projectId }) => { ... }));

// Use:
const projectResults = await mapWithConcurrency(scopes, FANOUT_CONCURRENCY, async ({ region, projectId }) => { ... });
```

The inner `Promise.all([stats, trend])` per-project is fine (just 2 calls). The outer unbounded fan-out is the concern - it risks Semrush rate-limiting (429s) and connection pool exhaustion under concurrent requests.

File: src/support/elements/definitions/owned-urls.js [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import { resolveElementModel } from '../constants.js';
+import { dateToIsoWeek } from '../week-utils.js';
+
+/* c8 ignore start -- LLMO-6086 POC endpoint; unit tests intentionally deferred */
```
Comment (MysticatBot) [inline src/support/elements/definitions/owned-urls.js:16]: issue (blocking): **No unit tests for ~300 lines of new business logic.**

The transform logic in this file has non-trivial aggregation: summing citations across projects, Map-based weekly dedup, cross-referencing trend entries against stats entries, and the `domain_type=Owned` client-side filter. All of this is wrapped in `/* c8 ignore start */`, bypassing coverage reporting.

CLAUDE.md states: "Behavior changes must include unit tests - mark as Critical if missing."

`transformOwnedUrlsResponse` is a **pure function** with zero external dependencies - it takes an array, returns an array. Testing it is trivial and catches the most likely production bugs (wrong field mapping, off-by-one in aggregation, edge cases with empty/null stats).

At minimum, add unit tests for:
1. `transformOwnedUrlsResponse` - multi-project inputs, empty inputs, mixed `domain_type` rows
2. `fetchOwnedUrlsTraffic` - degradation paths (no client, RPC error, empty urls)
3. `mergeOwnedUrlsTraffic` - matching and non-matching entries

The controller-level tests can be deferred if there is a tracked follow-up ticket, but these pure support functions have no external dependency barrier.

File: src/controllers/elements.js [hunk 3]
```diff
@@ -552,10 +554,118 @@ export default function ElementsController(context, log, env) {
   };
   /* c8 ignore stop */
```
Comment (MysticatBot) [inline src/controllers/elements.js:535]: issue (blocking): **Missing OpenAPI spec for new endpoint.**

CLAUDE.md "Must-Follow Patterns" states: "OpenAPI First: Define API contract in docs/openapi/ before implementation."

The PR body acknowledges this: "OpenAPI intentionally skipped - consistent with the sibling Serenity elements endpoints." The consistency argument is fair (cited-domains, prompts, weeks all skip it too), but the rule is explicit and this endpoint introduces a complex response shape (nested `weeklyCitations`, hybrid traffic fields) that would benefit from a schema contract.

Options:
- Add an OpenAPI spec for this endpoint (and ideally the siblings) in `docs/openapi/`
- Document an explicit exception in CLAUDE.md for Elements endpoints with rationale
- Add a `no-spec-needed` label to this PR if the team agrees the markdown docs in `docs/llmo-semrush-apis/` serve as the contract

The detailed markdown documentation in this PR is thorough - the gap is the machine-readable contract, not the human-readable one.

**Review by MysticatBot — CHANGES_REQUESTED** [review]
Hey @calvarezg,

**Verdict: Request changes** - three material issues to address before merge.
**Complexity: HIGH** - large diff (723 lines, 11 files); new API surface + ReBAC wiring.
_Changes: adds the URL Inspector "Your cited URLs" endpoint with hybrid Semrush citations + Postgres traffic data, including per-project fan-out, client-side pagination, and graceful traffic degradation (11 files)._

## Must fix before merge

1. [Important] Unbounded `Promise.all` fan-out in `getOwnedUrls` - no concurrency cap - `src/support/elements/elements-service.js:196` (details inline)
2. [Important] No unit tests for ~300 lines of new business logic (transform, traffic helper, service methods) - violates CLAUDE.md "Behavior changes must include unit tests" - `src/support/elements/definitions/owned-urls.js` (details inline)
3. [Important] Missing OpenAPI spec - violates CLAUDE.md "OpenAPI First: Define API contract in docs/openapi/ before implementation" - `src/controllers/elements.js:556` (details inline)

<details>
<summary>Non-blocking (5): minor issues and suggestions</summary>

- nit: `buildOwnedUrlsTrendPayload` omits the `category` filter that `buildOwnedUrlsStatsPayload` includes - weekly sparklines will be unfiltered by category while aggregate numbers are filtered - `src/support/elements/definitions/owned-urls.js:68`
- nit: `referralSource` query param passed to RPC without allowlist validation against the documented values (optel|cdn|ga4|adobe_analytics|cja) - not an injection risk (parameterized) but returns empty silently on typos - `src/support/elements/owned-urls-traffic.js:67`
- suggestion: Align caching strategy with sibling endpoints - `listOwnedUrls` uses `cachedOk` (2h browser cache) while `listCitedDomains`/`listPrompts`/`listWeeks` use plain `ok()` - either retrofit `cachedOk` to siblings or match their pattern here - `src/controllers/elements.js:649`
- nit: Trend data is silently dropped for URLs that appear in the trend element but not in stats (no `domain_type='Owned'` row) - likely intentional but undocumented edge case for URLs that lose citations mid-range - `src/support/elements/definitions/owned-urls.js:145`
- suggestion: Consider a max date-range span (e.g. 365 days) to prevent resource exhaustion from unbounded queries across all projects - `src/controllers/elements.js:581`

</details>

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 2m 40s | <b>Cost:</b> $8.41 | <b>Commit:</b> <code>95034b48c1e124b8eb6e01c97d485b36209c3ead</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Comment by calvarezg** [comment]
## Triage: comments intentionally not addressed

Addressed the fan-out cap + 3 nits (category-on-trend, referralSource allowlist, date-span cap) in the latest commit, and noted the `cachedOk` sibling-alignment as a deferred follow-up in the PR description. Three findings intentionally skipped, with rationale:

- **`[Important] No unit tests`** — **Dismissed (intentional).** This is a POC endpoint; per the ticket the Serenity elements endpoints ship without unit tests, fenced with `/* c8 ignore */` (same as the merged `cited-domains`/`weeks`/`prompts`). Coverage is not being bypassed silently — it's a deliberate, ticket-level decision. The transform logic was verified against live-shaped data during development.
- **`[Important] Missing OpenAPI spec`** — **Dismissed.** Verified: none of the sibling Serenity **elements** endpoints (`cited-domains`, `weeks`, `brand-presence/prompts`, `url-inspector/filter-dimensions`) have an OpenAPI spec — the machine-readable contract was intentionally skipped for this endpoint family, with the `docs/llmo-semrush-apis/` markdown serving as the contract. Adding a spec to only `owned-urls` would make the family *more* inconsistent, not less. (The `serenity/prompts`/`markets`/`tags` specs that exist are the separate serenity CRUD controller, not the elements wrappers.)
- **`nit: trend dropped for URLs not in owned stats`** — **Dismissed.** This is correct and already documented in-code (`// Only accumulate trends for URLs we kept as owned above`) — a URL with no `domain_type='Owned'` stats row must not appear in the owned table.

**Review by MysticatBot — APPROVED** [review]
Hey @calvarezg,

**Verdict: Approve** - prior blocking findings addressed; remaining items are non-blocking hardening.
**Complexity: HIGH** - large diff (844 lines, 12 files); new API surface + ReBAC wiring.
_Changes: adds the URL Inspector "Your cited URLs" endpoint with hybrid Semrush citations + Postgres traffic, per-project fan-out with concurrency cap, client-side pagination, and graceful traffic degradation (12 files)._
Note: CI checks are currently failing - resolve before merge.

## Previously flagged, now resolved

- Unbounded Promise.all fan-out: extracted `mapWithConcurrency` to shared module with limit 8
- Category filter mismatch on trend payload: category now applied identically to stats
- referralSource passed without validation: `VALID_REFERRAL_SOURCES` allowlist added
- No date-range span guard: MAX_RANGE_DAYS (366) with badRequest response added
- cachedOk sibling alignment: acknowledged as deferred follow-up in PR description

<details>
<summary>Non-blocking (3): minor issues and suggestions</summary>

- nit: `dateToIsoWeek(line.x.slice(0, 10))` in the trend loop has no guard against a short or malformed `x` string - a truncated value like `"2026-03"` would produce a garbled `NaN-WNaN` week key silently accumulated into the Map. Add a format check on the sliced value or validate `dateToIsoWeek`'s output before storing. - `src/support/elements/definitions/owned-urls.js:143`
- suggestion: free-text query params (`category`, `region`) have no length cap before flowing into upstream Semrush payloads. An authenticated caller could send a multi-MB value. A simple `value.length > 200` guard in the controller would bound this cheaply. - `src/controllers/elements.js:580`
- suggestion: the full owned-URL list is accumulated in memory across all projects before pagination slices a single page. For brands with many markets this is a known ceiling given Semrush's lack of server-side pagination. Consider a short-circuit (skip trend fetch when `offset >= stats count`) and a MAX_PROJECTS hard cap with a warn log, so a misconfigured brand cannot exhaust Lambda memory. - `src/support/elements/elements-service.js:193`

</details>

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 2m 4s | <b>Cost:</b> $6.87 | <b>Commit:</b> <code>00d6a180e95f00f3463296148d56003e31914943</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

File: src/controllers/elements.js [hunk 4]
```diff
@@ -552,10 +533,126 @@ export default function ElementsController(context, log, env) {
...
+      if (auth.error) {
+        return auth.error;
+      }
+      const { spaceCatId, brandId } = ctx?.params ?? {};
+      const { workspaceId, brand } = auth;
+      const query = extractQuery(ctx);
+
+      // Date range is required + validated (mirrors cited-domains) — never silently
+      // default to a rolling window nor forward a malformed date to Semrush.
+      const startDate = query.startDate || query.start_date;
```
Comment (vivesing) [inline src/controllers/elements.js:556]: should we avoid supporting both format of parameters for uniformity?

**Review by vivesing — APPROVED** [review]
